README.md
# Pace Calculator

This program is a pace calculator written in java. It will hopefully serve as a framework for future projects using the Java Swing forms available in Netbeans IDE. If you have any suggestions on ways to improve the project, feel free to open a feature request or an issue form to let me know. Right now, it is completely text based, but as mentioned above, there will be a GUI version to come in the future.

To use it, you can either download the repository and run it from  an ide such as Eclipse or Netbeans, or if you know how to compile it yourself, you can run the file from the terminal.


CHANGELOG.md
# CHANGELOG

### v0.0.1
Converted a set value to 1600, and 5000 meter times.

### v0.0.2
Added support for a much wider range of starting units and times.

### v0.0.3
Bug fixes

### v0.0.4
Added support for the 3200 meter distance as well as half and full marathon.

### v0.1.0
Pace Calculator now supports non metric conversions!

### v0.1.1
Increased efficiency with condensed methods that make use of parallel arrays.

### v0.1.2
Bug Fixes


PaceDriver.java
import java.util.Scanner;
public class PaceDriver {
	public static void main(String[] args) {
		Scanner scan = new Scanner(System.in);
		int distance, min, sec, tempInt;
		String tempStr;
		System.out.println("-------------");
		System.out.println("| 1. 1600   |");
		System.out.println("| 2. 3200   |");
		System.out.println("| 3. 5000   |");
		System.out.println("| 4. Custom |");
		System.out.println("-------------");
		System.out.print("Selection: ");
		distance =  scan.nextInt();
		if(distance == 4){
			System.out.println("-------------");
			System.out.println("| 1. Meters |");
			System.out.println("| 2. Miles  |");
			System.out.println("-------------");
			tempInt = scan.nextInt();
		}
		System.out.print("Time: ");
		tempStr = scan.nextLine();
		if(tempStr.length() == 4){
			min = Integer.parseInt(tempStr.substring(1,2));
			sec = Integer.parseInt(tempStr.substring(1,2));
		}
		else if(tempStr.length() == 5){
			min = Integer.parseInt(tempStr.substring(1,2));
			sec = Integer.parseInt(tempStr.substring(1,2));
		}
		else {
			System.out.println("Not a valid time");
			min = 0;
			sec = 0;
		}

		Time time1 = new Time(min, sec, distance);
	}
	
}


src/CalcTime.java
import java.util.Scanner;
public class CalcTime {
	public static void main(String[] args) {
		// Objects
		Scanner scan = new Scanner(System.in);
		// Variables
		int distance, min, sec, tempInt, distanceSel;
		int hr = 0;																				// In case not initialized
		boolean pushUnits = false, hours = false;
		String tempStr;
		char units = 'a';																	// Means nothing

		// Distance menu
		System.out.println("-------------");
		System.out.println("| 1. 1600m  |");
		System.out.println("| 2. 3200m  |");
		System.out.println("| 3. 5000m  |");
		System.out.println("| 4. 1 Mile |");
		System.out.println("| 5. 2 Mile |");
		System.out.println("| 6. 3 Mile |");
		System.out.println("| 7. Half   |");
		System.out.println("| 8. Full   |");
		System.out.println("| 9. Custom |");
		System.out.println("-------------");
		System.out.print("Selection: ");
		distanceSel =  scan.nextInt();
		// Custom distance
		if(distanceSel == 9){
			pushUnits = true;
			// Menu
			System.out.println("-------------");
			System.out.println("| 1. Meters |");
			System.out.println("| 2. Miles  |");
			System.out.println("-------------");
			System.out.print("Selection");
			tempInt = scan.nextInt();
			if(tempInt == 1){
				units = 'e';
			}
			else if(tempInt == 2){
				units = 'i';
			}
			else{
				// Error Catching
				while((tempInt!=1)&&(tempInt!=2)){
					System.out.println("Please select either 1 or 2.");
					System.out.print("Selection: ");
					tempInt = scan.nextInt();
				}
			}
			System.out.print("Distance: ");
			distance = scan.nextInt();
		}
		distance = Time.dSelection(distanceSel);
		tempStr = null;
		System.out.print("Time: ");
		tempStr = scan.next();
		scan.close();
		// Seperates given time into hours, minutes, and seconds
		if(tempStr.length() == 4){
			min = Integer.parseInt(tempStr.substring(0,1));
			sec = Integer.parseInt(tempStr.substring(2,4));
		}
		else if(tempStr.length() == 5){
			min = Integer.parseInt(tempStr.substring(0,2));
			sec = Integer.parseInt(tempStr.substring(3,5));
		}
		else if(tempStr.length() == 7){
			hr = Integer.parseInt(tempStr.substring(0,1));
			min = Integer.parseInt(tempStr.substring(2,4));
			sec = Integer.parseInt(tempStr.substring(5,7));
			hours = true;
		}
		else {
			// Error Catching
			System.out.println("Not a valid time");
			min = 0;
			sec = 0;
		}
		// Object Initialization
		Time time1 = new Time(min, sec, distance);
		// If custom distance
		if(pushUnits){
			time1.pushUnits(units);
		}
		// If marathon/half marathon
		if(hours){
			time1.pushHour(hr);
		}

		// Prints
		for(int i = 0; i<8; i++){
			System.out.println(time1.getTime(i));
		}
	}
}


src/Time.java
public class Time {
	// Instance data
	private char units;
	private int totalTime;
	private int totalDistance;
	private double distances[] = {1600.0, 3200.0, 5000.0, 1609.0, 3218.0, 4828.0, 21082.0, 42164.0};
	private String distanceS[] = {"1600", "3200", "5000", "1 Mile", "2 Mile", "3 Mile", "Half", "Full"};
	public Time(int min, int sec, int distance){
		totalTime = min*60+sec;
    totalDistance = distance;
	}
  private int calculate(int x){
		double num = distances[x]/(double)totalDistance;
    return (int)(Math.pow(num, 1.06)*totalTime);
	}
	public String getTime(int x){
		if((distanceS[x].equals("Half"))||(distanceS[x].equals("Full"))){
			int tempS = this.calculate(x) % 60;
	    int totalMinutes = this.calculate(x) / 60;
	    int tempM = totalMinutes % 60;
	    int tempH = totalMinutes / 60;
			if(tempS<9){
				if(tempM<9){
					return(distanceS[x]+": "+tempH+":0"+tempM+":0"+tempS);
				}
				else{
					return(distanceS[x]+": "+tempH+":"+tempM+":0"+tempS);
				}
			}
			else{
				if(tempM<9){
					return(distanceS[x]+": "+tempH+":0"+tempM+":"+tempS);
				}
				else{
					return(distanceS[x]+": "+tempH+":"+tempM+":"+tempS);
				}
			}
		}
		else{
			int tempM = this.calculate(x)/60;
			int tempS = this.calculate(x)%60;
			if(tempS<9){
				return(distanceS[x]+": "+tempM+":0"+tempS);
			}
			else{
				return (distanceS[x]+": "+tempM+":"+tempS);
			}
		}
	}
	public int getTime(){
		return totalTime;
	}
	public void pushUnits(char iOrE){
		// Miles = i, Meters = e;
		if((iOrE == 'i')||(iOrE =='e'))
			units = iOrE;
		else
			System.out.println("Error: Not a valid character");
	}
	public void pushHour(int hour){
		totalTime = totalTime+(hour*60*60);
	}
	public static int dSelection(int selection){
		int tempDistance;
		switch(selection){
			case 1:	tempDistance = 1600;
			break;
			case 2: tempDistance = 3200;
			break;
			case 3:	tempDistance = 5000;
			break;
			case 4:	tempDistance = 1609;		// One Mile
			break;
			case 5: tempDistance = 3218;		// Two Mile
			break;
			case 6: tempDistance = 4828;		// Three Mile
			break;
			case 7:	tempDistance = 21082;		// Half Marathon
			break;
			case 8:	tempDistance =  42164;	// Full Marathon
			break;
   		default:tempDistance = 5000;		// Defaults to a 5k for XC season
    	break;
		}
		return tempDistance;
	}
	public String toString(){
    int min = totalTime/60;
    int sec = totalTime%60;
    if(sec<9)
      return totalDistance+" | "+min+":0"+sec;
    else
      return totalDistance+" | "+min+":"+sec;
	}
}
